provider/vsphere/ova_utils.go

```diff
@@ -5,79 +5,165 @@ package vsphere
 
 import (
 	"archive/tar"
+	"crypto/sha256"
+	"fmt"
 	"io"
 	"io/ioutil"
 	"net/http"
 	"os"
 	"path/filepath"
 
 	"github.com/juju/errors"
+	"github.com/juju/mutex"
 )
 
-func downloadOva(basePath, url string) (string, error) {
-	logger.Debugf("Downloading ova file from url: %s", url)
-	resp, err := http.Get(url)
+// CacheLocker is an interface that may be used for locking
+// an OVA cache directory.
+type CacheLocker interface {
+	// Lock locks the OVA cache directory for concurrent access,
+	// and returns a function to release the lock or an error.
+	Lock() (func(), error)
+}
+
+type mutexCacheLocker struct {
+	spec mutex.Spec
+}
+
+// NewMutexCacheLocker returns an implemenation of CacheLocker
+// based on juju/mutex, using the given mutex.Spec.
+func NewMutexCacheLocker(spec mutex.Spec) CacheLocker {
+	return mutexCacheLocker{spec: spec}
+}
+
+// Lock is part of the CacheLocker interface.
+func (l mutexCacheLocker) Lock() (func(), error) {
+	releaser, err := mutex.Acquire(l.spec)
+	if err != nil {
+		return nil, err
+	}
+	return releaser.Release, nil
+}
+
+// downloadOVA downloads and extracts the OVA identified by the
+// given image metadata into the specified directory. We store
```

Comment:
> Is using a directory the best choice? Previously when we supported caching LXC images, we used the mongo blob store because in a HA scenario, any controller could service the request. Using a directory means each controller will have it's own cache. Why not cache the images in mongo?

Maybe not in terms of reducing the number of downloads globally, but in practice I think it's a fine trade-off. How would caching in mongo even work? There's two obvious options:
 - have all the logic that's in the vsphere provider in some apiserver facade, breaking all the abstractions and encapsulation
 - expose a complex object store interface to providers, along with leases to ensure agents don't step on each other

IMO, the first option is a terrible idea. Our abstractions today are not great, and that has and continues to hurt us; further smearing logic around like that won't help as any. The second idea would be fine, but is complex due to the tension between concurrency & caching. I don't think that complexity is worth the meagre benefit.

LXC was a bit different, because there were many machines *all* downloading the same image concurrently. Only one agent will run the cloud instance provisioner (i.e. it's a singular worker), so only one machine will downloading at a time.